### PR TITLE
feat[venom]: add algebraic optimization pass

### PR DESCRIPTION
**Pull request by harkal**
### What I did

Added a venom pass to do algebraic optimizations. Currently optimizes `iszero` chains.

### Commit message

```
Add a new venom pass to do algebraic optimizations. 

Currently optimizes `iszero` chains.
```

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: vyper/venom/passes/algebraic_optimization.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+                    continue
+
+                for use_inst in self.dfg.get_uses(inst.output):
+                    opcode = use_inst.opcode
+
+                    if opcode == "iszero":
+                        # We keep iszero instuctions as is
+                        continue
+                    if opcode in ("jnz",):
+                        # instructions that accept an integer as input:
```
Comment (charles-cooper) [inline vyper/venom/passes/algebraic_optimization.py:34]:
```suggestion
                        # instructions that accept a truthy value as input:
```

File: vyper/venom/passes/algebraic_optimization.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                if iszero_count == 0:
+                    continue
+
+                for use_inst in self.dfg.get_uses(inst.output):
+                    opcode = use_inst.opcode
+
+                    if opcode == "iszero":
+                        # We keep iszero instuctions as is
+                        continue
+                    if opcode in ("jnz",):
```
Comment (charles-cooper) [inline vyper/venom/passes/algebraic_optimization.py:33]: i think we can add `assert` here too right?

File: vyper/venom/passes/algebraic_optimization.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+        return chain
+
+    def run_pass(self):
+        self.dfg = self.analyses_cache.request_analysis(DFGAnalysis)
+
+        self._optimize_iszero_chains()
+
+        self.analyses_cache.invalidate_analysis(DFGAnalysis)
+        self.analyses_cache.invalidate_analysis(CFGAnalysis)
```
Comment (charles-cooper) [inline vyper/venom/passes/algebraic_optimization.py:66]: should invalidate LivenessAnalysis, not CFG right?

File: vyper/venom/passes/algebraic_optimization.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+    def _get_iszero_chain(self, op: IROperand) -> list[IRInstruction]:
+        chain: list[IRInstruction] = []
+
+        while True:
+            inst = self.dfg.get_producing_instruction(op)
+            if inst is None or inst.opcode != "iszero":
+                break
+            op = inst.operands[0]
+            chain.insert(0, inst)
```
Comment (charles-cooper) [inline vyper/venom/passes/algebraic_optimization.py:56]: `chain.append(inst)` followed by `chain.reverse()` at the end can avoid some performance blowups for long lists

File: vyper/venom/passes/algebraic_optimization.py (hunk 3 above)
Comment (harkal) [reply vyper/venom/passes/algebraic_optimization.py]: right!

File: tests/unit/compiler/venom/test_algebraic_optimizer.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    RemoveUnusedVariablesPass(ac, fn).run_pass()
+
+    assert bb.instructions[-1].opcode == "jnz"
+    if (interleave_point + iszeros_after_interleave_point) % 2 == 0:
+        assert bb.instructions[-1].operands[0] == op3_inv
+    else:
+        assert bb.instructions[-1].operands[0] == op3
+
+
+test_simple_bool_cast_case(5)
```
Comment (charles-cooper) [inline tests/unit/compiler/venom/test_algebraic_optimizer.py:132]: stray function call

**Review by charles-cooper — APPROVED** [review]
nice!

**Pull request merged**